File: java-mutation-testing/src/main/java/com/codebreeze/mutation/hotel/Hotel.java
package com.codebreeze.mutation.hotel;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.IntStream;

import static com.google.common.base.Preconditions.checkState;
import static com.google.common.collect.Sets.newHashSet;

public class Hotel {
    private final Set<Room> availableRooms = newHashSet();
    private final Set<Reservation> reservations = newHashSet();

    public Hotel(final int roomCount){
        checkState(roomCount > 0, "cannot have a hotel with negative or zero room count [%d]", roomCount);
        IntStream.range(0, roomCount).forEach(i -> availableRooms.add(createRoom(i)));
    }

    private Room createRoom(int i) {
        return new Room(String.valueOf(i));
    }

    public Reservation reserve(final String roomNumber,
                               final LocalDateTime from,
                               final LocalDateTime to) {
        checkState(isAvailable(roomNumber), "room [%s] is already reserved.", roomNumber);
        makeUnavailable(roomNumber);
        final Reservation reservation = new Reservation(roomNumber, from, to);
        reservations.add(reservation);
        return reservation;
    }

    public void cancel(final String reservationNumber){
        final Optional<Reservation> reservation = reservations
                .stream()
                .filter(hasReservationNumber(reservationNumber))
                .findFirst();
        reservation.ifPresent(r -> availableRooms.add(new Room(r.getRoomNumber())));
        final boolean removed = reservations.removeIf(hasReservationNumber(reservationNumber));
        checkState(removed, "reservation not found [%s]", reservationNumber);
    }

    private Predicate<Reservation> hasReservationNumber(String reservationNumber) {
        return r -> r.getNumber().equals(reservationNumber);
    }

    private boolean makeUnavailable(String roomNumber) {
        return availableRooms.removeIf(hasRoomNumber(roomNumber));
    }

    private boolean isAvailable(final String roomNumber) {
        return availableRooms.stream().anyMatch(hasRoomNumber(roomNumber));
    }

    private Predicate<Room> hasRoomNumber(String roomNumber) {
        return r -> r.getNumber().equals(roomNumber);
    }
}

File: javascript-mutation-testing/src/Hotel.js
var checkArgument = require('conditional').checkArgument;
var _ = require('underscore');

var room = function (number) {
    return Object.freeze({
        number: number
    });
};

var reservation = function (number, roomNumber, from, to) {
    return Object.freeze({
        number: number,
        roomNumber: roomNumber,
        from: from,
        to: to,
        creationTime: new Date()
    });
};

var hotel = function (roomCount) {
    console.log(roomCount);
    checkArgument((typeof roomCount === 'number') && roomCount > 0, "room count in a hotel must be positive integer");
    var makeRooms = function () {
        var rooms = [];
        _.each(_.range(0, roomCount), function (number) {
            rooms.push(room(number))
        });
        return rooms;
    };

    var availableRooms = makeRooms();
    var reservations = [];

    var hasNumber = function (number) {
        return function (r) {
            return r.number === number;
        };
    };

    var isAvailable = function (roomNumber) {
        return _.some(availableRooms, hasNumber(roomNumber));
    };

    var makeUnavailable = function (roomNumber) {
        availableRooms = _.reject(availableRooms, hasNumber(roomNumber))
    };

    var reserve = function (roomNumber, from, to) {
        checkArgument(isAvailable(roomNumber), "room is already reserved");
        makeUnavailable(roomNumber);
        var res = reservation('1', roomNumber, from, to);
        reservations.push(res);
        return res;
    };

    var cancel = function (reservationNumber) {
        var reservation = _.find(reservations, hasNumber(reservationNumber));
        checkArgument(reservation !== undefined, "reservation not found");
        reservations = _.reject(reservations, hasNumber(reservationNumber));
        availableRooms.push(Room(reservation.roomNumber));
    };

    return Object.freeze({
        cancel: cancel,
        reserve: reserve
    });
};

module.exports = hotel;